---
title: An introduction to contrast decomposition and querying using orthos
author: 
- name: Panagiotis Papasaikas
  affiliation: 
  - &id3 Friedrich Miescher Institute for Biomedical Research, Basel, Switzerland
  - SIB Swiss Institute of Bioinformatics
  email: panp80@gmail.com
- name: Charlotte Soneson
  affiliation: 
  - &id3 Friedrich Miescher Institute for Biomedical Research, Basel, Switzerland
  - SIB Swiss Institute of Bioinformatics
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('orthos')`"
output:
  BiocStyle::html_document:
    toc_float: true
editor_options: 
  chunk_output_type: console
bibliography: orthos.bib
geometry: "left=0cm,right=3cm,top=2cm,bottom=2cm"
vignette: >
  %\VignetteIndexEntry{1. Introduction to orthos}
  %\VignetteEncoding{UTF-8}
  %\VignettePackage{orthos}
  %\VignetteKeywords{cVAE, VariationalAutoEncoders, DGE, DifferentialGeneExpression, RNASeq}    
  %\VignetteEngine{knitr::rmarkdown}
---


```{r setup, include = FALSE, echo=FALSE, results="hide", message=FALSE}
require(knitr)

knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    error = FALSE,
    warning = FALSE,
    message = FALSE,
    crop = NULL
)
stopifnot(requireNamespace("htmltools"))
htmltools::tagList(rmarkdown::html_dependency_font_awesome())
```


```{r, echo=FALSE, out.width='50%', fig.align='center'}
knitr::include_graphics(path = system.file(package="orthos", "orthos.png", mustWork=TRUE))
```


```{css, echo = FALSE}
body {
  margin: 0 auto;
  max-width: 1600px;
  padding: 2rem;
}
```










# Introduction

Differential expression analysis upon cellular perturbations is the most commonly used tool for probing molecular mechanisms of action due to its simplicity and low cost. 
However, the interpretation of such gene expression contrasts is confounded by the complex and nuanced impact of experimental treatments on cellular processes. 

`orthos` is a generative modelling-based approach that disentangles the experiment-specific from the non-specific effects of perturbations on gene expression. 
The model is trained on a large corpus (>60K annotated / >0.5M augmented per organism) of gene expression contrasts compiled from the [ARCHS4](https://maayanlab.cloud/archs4/) database of uniformly processed RNAseq experiments. 
It accurately captures and isolates "**polytopic**" (non-specific, observed across multiple treatments) effects while accounting for context. 
The residual, **specific** component obtained from this decomposition offers a more unequivocal experimental signature and is a proxy more closely related to the direct molecular effects of a treatment when compared to the raw signal.

In addition to providing a clearer understanding of the effects of experimental treatments on gene expression, `orthos` also enables researchers to query the contrast database with arbitrary contrasts and identify experiments with similar specific effects, ultimately helping to map treatments to mechanisms of action.

After installation, the package can be loaded to the user workspace with:

```{r library}
library(orthos)
```


A typical analysis involves two steps:

1. Decomposing a contrast/set of contrasts into the(ir) corresponding specific and polytopic fractions using the `decomposeVar()` function and

2. Performing a query with the original and decomposed (specific, polytopic) fractions against the corresponding contrast DBs using the `queryWithContrasts()` function.




# Demonstration data

To demonstrate the functionality of orthos we use a dataset from the the GEO submission [GSE215150](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE215150).

This series was not part of the `orthos` training or the `orthosDB` constrast database as it was publicly released on January 1st 2023 after the freeze of the training data to the ARCHS4 v2.1.2 database.

The performed experiment involves overexpression (OE) of the MKL/megakaryoblastic leukemia 1 gene (also termed MRTFA/myocardin related transcription factor A) and a constitutively active mutant MKL1 (caMKL1, described in @hu2019mkl1). 
Both OEs were performed in mouse LM2 and human 4T1 tumor-derived breast cancer cell lines. 
In addition to the MKL1/caMKL1 OE treatments the series also contains no-treatment controls for each of the two cell lines.

For simplicity the three biological replicates of each condition available in the original series have been collapsed in the data provided in the package.

In the provided form each of the two datasets (Mouse, Human) contains raw counts for >55000 genes identified by gene symbols and three conditions, namely, Control (Ctrl), MKL1 OE (MKL1) and constitutive MKL1 OE (caMKL1):

```{r}
MKL1_human <- readRDS(system.file("data", "GSE215150_MKL1_Human.rds", package = "orthos"))
head(MKL1_human)
dim(MKL1_human)
MKL1_mouse <- readRDS(system.file("data", "GSE215150_MKL1_Mouse.rds", package = "orthos"))
head(MKL1_mouse)
dim(MKL1_mouse)
```





# Decomposition of differential gene expression variance into specific and non-specific fractions using `decomposeVar()`

The workhorse behind `orthos` are organism-specific conditional variational autoencoder (cVAE) models that break down the variance of a given differential expression experiment into a non-specific (polytopic) and an experiment-specific fraction. 
The former corresponds to gene variance that has been observed across multiple instances during training while the latter is variance that is fairly unique to the experiment. 

The inputs to the models are gene counts in the form of log-transformed counts-per-million (lcpms) that are used to encode the **"context"** of the performed experiment as well as the actual gene expression contrasts in the form of gene log fold ratios (LFCs). 
As we will see, calculation of those inputs is by default performed internally given only **raw gene counts** and a specification of the contrasted conditions.

Given these inputs the model maps the contrast to a concise latent representation (z~D~) which is then used to generate a decoded version of the contrast. 
This decoded contrast corresponds directly to the **polytopic** (i.e non-specific) fraction of the observed effects; it subsumes the gene variance that the model can account for because it has been encountered, in some form, during training. 
The residual obtained after removing the decoded contrast from the original one is the specific fraction; this encompasses the gene variance component that the model cannot account for (random noise + specific biological effects).  

![orthos architecture](orthos_arch.png){width=100%}

The function that performs this contrast decomposition into the polytopic and specific fractions is `decomposeVar()`.
There are two available modes in which the required inputs can be fed into the function:

- In the first mode the user passes the matrix `M` of raw counts (genes in rows, conditions in columns) and two vectors `treatm` and `cntr` of equal length specifying the `M` column indices corresponding to the "treatments" and their respective "controls". 
The same column indices can be repeated multiple times in these vectors as for example in the case where multiple treatments are paired to the same control. 

- In the second mode the user passes the matrix `M` of raw counts and a second matrix `MD` that contains precalculated LFCs for the contrasts to be analyzed. 
In this case `M` specifies the context for the corresponding columns of `MD`. 
Naturally in this case the two matrices need to have the same dimensionality and matching row and column names. 

In both modes the rownames of `M` (and `MD` if specified) need to correspond to some valid gene identifiers (Entrez gene ids, ENSEMBL gene ids, gene symbols or ARCHS4 gene ids). 
By default the type of gene identifier used is detected automatically.


For the MKL1 data typical calls of `decomposeVar()` using mode 1 would look like this:

```{r, warning = FALSE, message = FALSE}
dec_MKL1_human <- decomposeVar(M = MKL1_human, treatm = c(2, 3), cntr = c(1, 1), 
                               organism = "Human", verbose = FALSE)

dec_MKL1_mouse <- decomposeVar(M = MKL1_mouse, treatm = c(2, 3), cntr = c(1, 1),
                               organism = "Mouse", verbose = FALSE)
```

Notice that, of the total gene features present in the input (>55K), only those that are part of the `orthos` model (~20K genes) are used.
These ~20K `orthos` genes are "sanctioned" according to several criteria (located on canonical chromosomes, no pseudogenes, no ribosomal protein genes, detected in at least a small fraction of the ARCHS4 database). 

The model is highly robust to small fractions of `orthos` genes not being part of the user input, even if those genes are expressed in the context under consideration. 
That being noted, it is safer to feed in inputs that are as comprehensive as possible (i.e not filtered in any way) in terms of gene features.






The output of the function is a `SummarizedExperiment` object with dimensions `N` x `M` where `N` is the number of `orthos` genes for that organism and `M` is the number of contrasts specified during input. 
The object has 4 assay slots corresponding to the input contrasts, decoded (i.e polytopic), and residual (i.e specific) fractions as well as the gene context. 
Contrasts are represented as LFCs and context is represented as log transformed cpms (lcpms).  
The `colData` of the object summarizes the proportion of variance accounted for in each decomposed fraction:

```{r}
summary(dec_MKL1_human)
colData(dec_MKL1_human) 

summary(dec_MKL1_mouse)
colData(dec_MKL1_mouse)
```



# Querying the database of gene contrasts using  `queryWithContrasts()`

Typically the next step of the analysis involves querying the contrasts database (orthosData) to identify public experiments similar to the one(s) under investigation in terms of the original, or decomposed (polytopic, specific fractions). 
orthosData contains >100K DGE experiments compiled from the ARCHS4 database of publicly available expression data. 
Each entry in orthosData corresponds to a pair of RNAseq samples contrasting a treatment vs a control. 
A combination of metadata semantic and quantitative analysis was used to determine the proper assignment of samples to such pairs in orthosData.

The function that performs the queries of user-provided decomposed samples against orthosData is `queryWithContrasts()`. 
The input to this function is the `SummarizedExperiment` object obtained in the previous step from `decomposeVar()`. 
However, users can use a column-subsetted version of this object in case they wish to query the DB with a subset of their decomposed datasets.  

The function returns a list with three elements per dataset:

- Element "pearson.rhos", is itself a list with each element containing the Pearson correlation values against all the orthosData entries for a specific fraction (input, decoded (polytopic), residual (specific)). 
- Element "zscores", is also a list with each element containing the z-score transformed version of the values above.
- Element "TopHits", also a list with detailed orthosData metadata for each of the top `detailTopn` hits per fraction (default 10).


In the following examples please note that the queries are run in "DEMO" mode for the sake of efficiency. 
For actual analysis purposes, the default "ANALYSIS" mode should always be used.

```{r, fig.width=12}
# Parallelization parameters:
BPPARAM <- BiocParallel::MulticoreParam(workers=2)

#For demonstration purposes. For actual analyses 'mode' should always be "ANALYSIS" (default):
query.res.human <- queryWithContrasts(dec_MKL1_human, organism = "Human", 
                                      BPPARAM = BPPARAM, verbose = FALSE, 
                                      mode = "DEMO")

names(query.res.human)

names(query.res.human$zscores)

# Query datasets in rows, orthosData entries in columns:
dim(query.res.human$zscores$RESIDUAL_CONTRASTS) 
summary(t(query.res.human$zscores$RESIDUAL_CONTRASTS))

# Query the database using only the "caMKL1" mouse contrast, suppress plotting:
# For demonstration purposes. For actual analyses 'mode' should always be "ANALYSIS" (default):
query.res.mouse <- queryWithContrasts(dec_MKL1_mouse[, "caMKL1"], organism = "Mouse", 
                                      BPPARAM = BPPARAM, verbose = FALSE, 
                                      plotType = "none", mode = "DEMO")

#Plot the query results for individual contrasts using violin plots:
ViolinPlots_mouse <- plotQueryResultsViolin(query.res.mouse, plot = FALSE)
ViolinPlots_mouse[["caMKL1"]]

#Plot the query results for individual contrasts using composite Manhattan/Density plots:
ManhDensPlots_mouse <- plotQueryResultsManh(query.res.mouse, plot = FALSE)
ManhDensPlots_mouse[["caMKL1"]]
```

# Accessing the contrast DB

!!! This is a placeholder, text and code to be finalized when orthosData are in place !!!

```{r}
organism <- "Mouse"
orthosDB <- HDF5Array::loadHDF5SummarizedExperiment(
    dir = "/tungstenfs/groups/gbioinfo/papapana/DEEP_LEARNING/Autoencoders/ARCHS4/Rdata/DECOMPOSED_CONTRASTS_HDF5",
    prefix = paste0(tolower(organism), "_v212_NDF_c100_DEMO")
)
par(mfrow = c(2,2))

queryID <- "GSM5021181"
for (contrast in names(assays(dec_MKL1_mouse))[c(4, 1, 2, 3)]) {
    plot(assays(dec_MKL1_mouse)[[contrast]][, "caMKL1"],
         assays(orthosDB)[[contrast]][, queryID],
         pch = 16, cex = 0.5, col = "darkslategrey", main = contrast,
         xlab = "LFC caMKL1", ylab = paste0("LFC ", queryID))
    abline(0, 1, col = "darkred", lwd = 0.8, lty = 2)
}     

```

# Accessing the model internals and other miscellaneous use cases

Producing latent encodings of contrasts
Producing latent encodings of contexts
Morphing decoded contrasts according to a different context
Fine-tuning the model to custom contrast datasets


# Session information {-}

```{r}
sessionInfo()
```

# References
